File: cluster-cli/src/main/scala/org/midonet/cluster/cli/commands/objects/Condition.scala
/*
 * Copyright 2015 Midokura SARL
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.midonet.cluster.cli.commands.objects

import java.util.UUID

import org.midonet.cluster.models.Topology.{Rule => TopologyRule}
import org.midonet.cluster.data.{ZoomClass, ZoomField, ZoomObject}
import org.midonet.cluster.util.IPSubnetUtil.{Converter => IPSubnetConverter}
import org.midonet.cluster.util.MACUtil.{Converter => MACConverter}
import org.midonet.cluster.util.RangeUtil.{Converter => RangeConverter}
import org.midonet.cluster.util.UUIDUtil.{Converter => UUIDConverter}
import org.midonet.packets.{IPSubnet, MAC}
import org.midonet.util.Range

/** A rule condition. */
@ZoomClass(clazz = classOf[TopologyRule])
@CliName(name = "condition", prepend = false)
class Condition extends ZoomObject with Obj {
    @ZoomField(name = "conjunction_inv")
    @CliName(name = "inv-condition", readonly = false)
    var invCondition: Boolean = _
    @ZoomField(name = "match_forward_flow")
    @CliName(name = "match-forward-flow", readonly = false)
    var matchForwardFlow: Boolean = _
    @ZoomField(name = "match_return_flow")
    @CliName(name = "match-return-flow", readonly = false)
    var matchReturnFlow: Boolean = _
    @ZoomField(name = "in_port_ids", converter = classOf[UUIDConverter])
    @CliName(name = "in-port-ids")
    var inPortIds: Set[UUID] = _
    @ZoomField(name = "in_port_inv")
    @CliName(name = "inv-in-port", readonly = false)
    var invInPort: Boolean = _
    @ZoomField(name = "out_port_ids", converter = classOf[UUIDConverter])
    @CliName(name = "out-port-ids")
    var outPortIds: Set[UUID] = _
    @ZoomField(name = "out_port_inv")
    @CliName(name = "inv-out-port", readonly = false)
    var invOutPort: Boolean = _
    @ZoomField(name = "port_group_id", converter = classOf[UUIDConverter])
    @CliName(name = "port-group-id", readonly = false)
    var portGroup: UUID = _
    @ZoomField(name = "inv_port_group")
    @CliName(name = "inv-port-group", readonly = false)
    var invPortGroup: Boolean = _
    @ZoomField(name = "ip_addr_group_id_src", converter = classOf[UUIDConverter])
    @CliName(name = "ip-addr-group-id-src", readonly = false)
    var ipAddrGroupIdSrc: UUID = _
    @ZoomField(name = "inv_ip_addr_group_id_src")
    @CliName(name = "inv-ip-addr-group-id-src", readonly = false)
    var invIpAddrGroupIdSrc: Boolean = _
    @ZoomField(name = "ip_addr_group_id_dst", converter = classOf[UUIDConverter])
    @CliName(name = "ip-addr-group-id-dst", readonly = false)
    var ipAddrGroupIdDst: UUID = _
    @ZoomField(name = "inv_ip_addr_group_id_dst")
    @CliName(name = "inv-ip-addr-group-id-dst", readonly = false)
    var invIpAddrGroupIdDst: Boolean = _
    @ZoomField(name = "dl_type")
    @CliName(name = "dl-type", readonly = false)
    var dlType: Int = _
    @ZoomField(name = "inv_dl_type")
    @CliName(name = "inv-dl-type", readonly = false)
    var invDlType: Boolean = _
    @ZoomField(name = "dl_src", converter = classOf[MACConverter])
    @CliName(name = "dl-src", readonly = false)
    var dlSrc: MAC = _
    @ZoomField(name = "dl_src_mask")
    @CliName(name = "dl-src-mask", readonly = false)
    var dlSrcMask: Long = -1L
    @ZoomField(name = "inv_dl_src")
    @CliName(name = "inv-dl-src", readonly = false)
    var invDlSrc: Boolean = _
    @ZoomField(name = "dl_dst", converter = classOf[MACConverter])
    @CliName(name = "dl-dst", readonly = false)
    var dlDst: MAC = _
    @ZoomField(name = "dl_dst_mask")
    @CliName(name = "dl-dst-mask", readonly = false)
    var dlDstMask: Long = -1L
    @ZoomField(name = "inv_dl_dst")
    @CliName(name = "inv-dl-dst", readonly = false)
    var invDlDst: Boolean = _
    @ZoomField(name = "nw_tos")
    @CliName(name = "nw-tos", readonly = false)
    var nwTos: Byte = _
    @ZoomField(name = "nw_tos_inv")
    @CliName(name = "inv-nw-tos", readonly = false)
    var invNwTos: Boolean = _
    @ZoomField(name = "nw_proto")
    @CliName(name = "nw-proto", readonly = false)
    var nwProto: Byte = _
    @ZoomField(name = "nw_proto_inv")
    @CliName(name = "inv-nw-proto", readonly = false)
    var invNwProto: Boolean = _
    @ZoomField(name = "nw_src_ip", converter = classOf[IPSubnetConverter])
    @CliName(name = "nw-src-ip", readonly = false)
    var nwSrcIp: IPSubnet[_] = _
    @ZoomField(name = "nw_src_inv")
    @CliName(name = "inv-nw-src", readonly = false)
    var invNwSrc: Boolean = _
    @ZoomField(name = "nw_dst_ip", converter = classOf[IPSubnetConverter])
    @CliName(name = "nw-dst-ip", readonly = false)
    var nwDstIp: IPSubnet[_] = _
    @ZoomField(name = "nw_dst_inv")
    @CliName(name = "inv-nw-dst", readonly = false)
    var invNwDst: Boolean = _
    @ZoomField(name = "tp_src", converter = classOf[RangeConverter])
    @CliName(name = "tp-src", readonly = false)
    var tpSrc: Range[Integer] = _
    @ZoomField(name = "tp_src_inv")
    @CliName(name = "inv-tp-src", readonly = false)
    var invTpSrc: Boolean = _
    @ZoomField(name = "tp_dst", converter = classOf[RangeConverter])
    @CliName(name = "tp-dst", readonly = false)
    var tpDst: Range[Integer] = _
    @ZoomField(name = "tp_dst_inv")
    @CliName(name = "inv-tp-dst", readonly = false)
    var invTpDst: Boolean = _
    @ZoomField(name = "traversed_device", converter = classOf[UUIDConverter])
    @CliName(name = "traversed-device", readonly = false)
    var traversedDevice: UUID = _
    @ZoomField(name = "traversed_device_inv")
    @CliName(name = "inv-traversed-device", readonly = false)
    var invTraversedDevice: Boolean = _
    @ZoomField(name = "fragment_policy")
    @CliName(name = "fragment-policy", readonly = false)
    var fragmentPolicy: FragmentPolicy = FragmentPolicy.UNFRAGMENTED
}

File: cluster/midonet-cluster/src/main/java/org/midonet/cluster/rest_api/models/Condition.java
/*
 * Copyright 2015 Midokura SARL
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.midonet.cluster.rest_api.models;

import java.lang.reflect.Type;
import java.net.URI;
import java.util.UUID;

import javax.validation.constraints.Max;
import javax.validation.constraints.Min;
import javax.validation.constraints.Pattern;
import javax.xml.bind.annotation.XmlTransient;

import com.google.protobuf.Message;

import org.codehaus.jackson.annotate.JsonIgnore;

import org.midonet.cluster.data.ZoomConvert;
import org.midonet.cluster.data.ZoomEnum;
import org.midonet.cluster.data.ZoomEnumValue;
import org.midonet.cluster.data.ZoomField;
import org.midonet.cluster.models.Topology;
import org.midonet.cluster.rest_api.BadRequestHttpException;
import org.midonet.cluster.rest_api.annotation.JsonError;
import org.midonet.cluster.rest_api.validation.IsValidFragmentType;
import org.midonet.cluster.rest_api.validation.ValidMac;
import org.midonet.cluster.rest_api.validation.MessageProperty;
import org.midonet.cluster.rest_api.validation.ValidMacMask;
import org.midonet.cluster.util.IPSubnetUtil;
import org.midonet.cluster.util.RangeUtil;
import org.midonet.cluster.util.UUIDUtil;
import org.midonet.packets.IPSubnet;
import org.midonet.packets.IPv4;
import org.midonet.packets.IPv4Subnet;
import org.midonet.packets.MAC;
import org.midonet.util.Range;
import org.midonet.util.version.Since;

@IsValidFragmentType
public class Condition extends UriResource {

    /** This enumeration is equivalent with
     * [[org.midonet.midolman.rules.FragmentPolicy]] except that this class
     * uses the lower-case values for API compatibility. */
    @ZoomEnum(clazz = Topology.Rule.FragmentPolicy.class)
    public enum FragmentPolicy {
        @ZoomEnumValue(value = "ANY") any,
        @ZoomEnumValue(value = "NONHEADER") nonheader,
        @ZoomEnumValue(value = "HEADER") header,
        @ZoomEnumValue(value = "UNFRAGMENTED") unfragmented
    }

    @XmlTransient
    @ZoomField(name = "nw_dst_ip", converter = IPSubnetUtil.Converter.class)
    public IPSubnet<?> nwDst;
    @Pattern(regexp = IPv4.regex, message = "is an invalid IP format")
    public String nwDstAddress;
    @Min(0)
    @Max(32)
    public int nwDstLength;
    @XmlTransient
    @ZoomField(name = "nw_src_ip", converter = IPSubnetUtil.Converter.class)
    public IPSubnet<?> nwSrc;
    @Pattern(regexp = IPv4.regex, message = "is an invalid IP format")
    public String nwSrcAddress;
    @Min(0)
    @Max(32)
    public int nwSrcLength;

    @ZoomField(name = "conjunction_inv")
    public boolean condInvert;
    @ZoomField(name = "match_forward_flow")
    public boolean matchForwardFlow;
    @ZoomField(name = "match_return_flow")
    public boolean matchReturnFlow;

    @ZoomField(name = "in_port_ids", converter = UUIDUtil.Converter.class)
    public UUID[] inPorts;
    @ZoomField(name = "in_port_inv")
    public boolean invInPorts;
    @ZoomField(name = "out_port_ids", converter = UUIDUtil.Converter.class)
    public UUID[] outPorts;
    @ZoomField(name = "out_port_inv")
    public boolean invOutPorts;
    @ZoomField(name = "port_group_id", converter = UUIDUtil.Converter.class)
    public UUID portGroup;
    @ZoomField(name = "inv_port_group")
    public boolean invPortGroup;
    @ZoomField(name = "ip_addr_group_id_src", converter = UUIDUtil.Converter.class)
    public UUID ipAddrGroupSrc;
    @ZoomField(name = "inv_ip_addr_group_id_src")
    public boolean invIpAddrGroupSrc;
    @ZoomField(name = "ip_addr_group_id_dst", converter = UUIDUtil.Converter.class)
    public UUID ipAddrGroupDst;
    @ZoomField(name = "inv_ip_addr_group_id_dst")
    public boolean invIpAddrGroupDst;

    @ZoomField(name = "traversed_device", converter = UUIDUtil.Converter.class)
    public UUID traversedDevice;
    @ZoomField(name = "traversed_device_inv")
    public boolean invTraversedDevice;

    @Min(0x0600)
    @Max(0xFFFF)
    @ZoomField(name = "dl_type")
    public Integer dlType;
    @ZoomField(name = "inv_dl_type")
    public boolean invDlType;

    @ValidMac
    @ZoomField(name = "dl_src")
    public String dlSrc;

    @Since("2")
    @ValidMacMask
    @ZoomField(name = "dl_src_mask", converter = MACMaskConverter.class)
    public String dlSrcMask;

    @ZoomField(name = "inv_dl_src")
    public boolean invDlSrc;

    @ZoomField(name = "dl_dst")
    @ValidMac
    public String dlDst;

    @Since("2")
    @ZoomField(name = "dl_dst_mask", converter = MACMaskConverter.class)
    @ValidMacMask
    public String dlDstMask;

    @ZoomField(name = "inv_dl_dst")
    public boolean invDlDst;
    @ZoomField(name = "nw_tos")
    public Integer nwTos;
    @ZoomField(name = "nw_tos_inv")
    public boolean invNwTos;
    @ZoomField(name = "nw_proto")
    public Integer nwProto;
    @ZoomField(name = "nw_proto_inv")
    public boolean invNwProto;
    @ZoomField(name = "nw_src_inv")
    public boolean invNwSrc;
    @ZoomField(name = "nw_dst_inv")
    public boolean invNwDst;

    @Since("2")
    @JsonError(message = MessageProperty.FRAG_POLICY_UNDEFINED)
    @ZoomField(name = "fragment_policy")
    public FragmentPolicy fragmentPolicy;

    @ZoomField(name = "tp_src", converter = RangeUtil.Converter.class)
    public Range<Integer> tpSrc;
    @ZoomField(name = "tp_dst", converter = RangeUtil.Converter.class)
    public Range<Integer> tpDst;

    @ZoomField(name = "tp_src_inv")
    public boolean invTpSrc;
    @ZoomField(name = "tp_dst_inv")
    public boolean invTpDst;

    @JsonIgnore
    @Override
    public void afterFromProto(Message proto) {
        nwDstAddress = nwDst != null ? nwDst.getAddress().toString() : null;
        nwDstLength = nwDst != null ? nwDst.getPrefixLen() : 0;
        nwSrcAddress = nwSrc != null ? nwSrc.getAddress().toString() : null;
        nwSrcLength = nwSrc != null ? nwSrc.getPrefixLen() : 0;
    }

    @JsonIgnore
    @Override
    public void beforeToProto() {
        nwDst = nwDstAddress != null ?
                IPv4Subnet.fromCidr(nwDstAddress + "/" + nwDstLength) : null;
        nwSrc = nwSrcAddress != null ?
                IPv4Subnet.fromCidr(nwSrcAddress + "/" + nwSrcLength) : null;
        fragmentPolicy = getAndValidateFragmentPolicy();
    }

    public boolean hasL4Fields() {
        return null != tpDst || null != tpSrc;
    }

    private FragmentPolicy getAndValidateFragmentPolicy() {
        if (null == fragmentPolicy)
            return hasL4Fields() ? FragmentPolicy.header : FragmentPolicy.any;

        if (hasL4Fields() &&
            (fragmentPolicy == FragmentPolicy.any ||
             fragmentPolicy == FragmentPolicy.nonheader)) {
            throw new BadRequestHttpException(MessageProperty.getMessage(
                MessageProperty.FRAG_POLICY_INVALID_FOR_L4_RULE));
        }

        return fragmentPolicy;
    }

    public static class MACMaskConverter
        extends ZoomConvert.Converter<String, Long> {
        @Override
        public String fromProto(Long value, Type clazz) {
            return MAC.maskToString(value);
        }
        @Override
        public Long toProto(String value, Type clazz) {
            return MAC.parseMask(value);
        }
    }

    public URI getUri() {
        // not in love with this model, but this is how it was done before.
        return null;
    }


}
